### ISSUE TITLE
Trying to understand the weakmap

### ISSUE THREAD

**Issue by lukiano** [metcoder95/fastify-racing#7]
Sorry for raising an issue with a question, couldn't find any other communication means. I'm trying to understand why this is using a WeakMap to store the controllers instead of storing them in Fastify's request context.

**Comment by metcoder95** [comment]
Hey! No worries at all.

>why this is using a WeakMap to store the controllers instead of storing them in Fastify's request context.

TL;DR: _to avoid adding overhead_.

The main idea was being able to create disposable `AbortControllers` that can be easily claimed by the GC in case of something blocks the execution of the `onResponse` hook (e.g. some other plugin blocks this, or something happened in the middle that the hook was never triggered).
By doing this we ensure that as soon as the `request` object is out of the scope and claimed by the GC, the Controller attached is also claimed to avoid holding memory just for nothing.

Along with that, the controller as it was might have some other overlaps with user-land decorations to the `request` object. So abstracting this and having the chance to claim the same controller over and over by just calling `req.race` all the time, easiest the DX meanwhile keeping the overhead as low as possible.

Hope it answers your question 🙂
